added: cli option --dump-config prints loaded config and exits

wolfgangwalther opened the pull request:
As discussed in https://github.com/PostgREST/postgrest/pull/1659#discussion_r528413899, this adds the CLI option `--dump-config`, which prints the config after loading and then exits. The printed config is itself a valid postgrest config file.

This is mostly useful as a base for future improvements. When adding aliases for some config options (https://github.com/PostgREST/postgrest/pull/1668#issuecomment-732367387) or reading options directly from environment variables (#1624, #1572), this can be used as the base to write proper tests.

Also added colors to our io-tests. They are all nice and green now. Much easier to spot the occasional red one...

wolfgangwalther commented on test/io-tests.sh:
```diff
@@ -36,17 +36,71 @@ trap cleanup sigint sigterm exit
 comment(){ echo "# $1"; }
 
+####################
+# FOREGROUND TESTS #
```
appreciate suggestions for a better name for this section and the "background" section. The distinction here is, that for those tests `postgrest` immediately returns and is not sent to the background.

Maybe it's just not needed at all. But it does give a bit of structure to that file. I'm thinking about running shellcheck against it.... but I am bit afraid of the result as well :D

steve-chavez replied:
Hmm.. I also don't have a better name, maybe just put the distinction as a comment?

wolfgangwalther commented on test/io-tests/configs/no-defaults.config:
```diff
@@ -0,0 +1,26 @@
+server-unix-socket-mode = "777"
+openapi-server-proxy-uri = "https://postgrest.org"
+jwt-secret = "c2VjdXJpdHl0aHJvdWdob2JzY3VyaXR5"
+jwt-aud = "postgrest.org"
```
I tried using `https://postgrest.org` here, because I understood https://tools.ietf.org/html/rfc7519#section-4.1.3 (that's what we are linking in the docs) to support either strings or URIs. But this results in an error when starting postgrest, because the `Crypto.JWT` parser for `StringOrUri` tells me, that this is not a valid URI.

How do I add a URI here? If somebody knows how, we could add a hint to the docs. If not for anyone else, it would at least help me :).

steve-chavez replied:
Weird.. I'm trying the `jwt-aud="https://postgrest.org"` now(on my set-configs branch) and it works.

steve-chavez replied:
Though there doesn't seem to be any test with an URI.. I've also tried `jwt-aud="//"` and it doesn't fail. Maybe the aud is not being validated at all..

wolfgangwalther replied:
I know that it's only treated as URI, if there is a `:` in it. And then it must be valid. I will try again with a valid and an invalid uri.

wolfgangwalther replied:
Or at least that's what I thought... I can confirm, any valid and invalid URL works right now.

steve-chavez replied:
I think we can open a new issue for this one.

wolfgangwalther commented:
Lot's of haskell again, @steve-chavez. Looking forward to your feedback! I already learned a bit from `hlint` :D.

wolfgangwalther commented:
Cirrus seems to be hitting a build timeout of 15 minutes now, saw that for a couple of commits in different PRs now. @steve-chavez you added a timeout of 30 minutes a little time ago. Did they lower the timeout or what's happening?

monacoremo commented on src/PostgREST/Config.hs:
```diff
@@ -62,9 +65,13 @@ import PostgREST.Types            (JSPath, JSPathExp (..),
+-- | Command line interface options
+data CLI = CLI
+  { cliDumpConfig :: Bool
+  , cliPath       :: FilePath }
```
As we are hoing to add options here that are mutually exclusive (e.g. dump schema cache), can we make this s sym type instead of bool?

monacoremo replied:
E.g. `data CliMode = Run | DumpConfig` (`| DumpSchemaCache` etc.)

wolfgangwalther replied:
Very good point, Remo. I think `--dump-config` as an option is wrong to start with, when those things are mutually exclusive.

Two ideas to make this better:
- Use `--dump=config` and `--dump=schema` - this is mutually exclusive from the start. Then have a `CliDump = DumpConfig | DumpSchema | DumpNothing`.
- Use a "sub command" based approach. So have subcommands like `postgrest run`, `postgrest dump-config` and `postgrest dump-schema`. Of course the default without command would be `run`. Those commands would be mutually exclusive as well by design. Then have `CliCmd = CmdRun | CmdDumpConfig | CmdDumpSchema` like you suggested. This would be a bit more flexible in case we're adding other commands in the future. Not sure how easy the "default to run without command" is with the cli parser.

I think I like the `--dump=...` option better. WDYT?

monacoremo replied:
--dump-config if good I think! Most straight forward. Optparse applicative will be able to enforce a consistent command (see 'alternatives' where a sum type is the basis and 'flags' in https://github.com/pcapriotti/optparse-applicative)

wolfgangwalther replied:
Ok, I went with a `--dump-config` flag and `data Command = CmdRun | CmdDumpConfig` for now.

steve-chavez commented:
> Cirrus seems to be hitting a build timeout of 15 minutes now, saw that for a couple of commits in different PRs now

@wolfgangwalther Hmm.. are you sure it's a timeout? It looks like it's failing because of low memory: https://github.com/haskell/cabal/issues/5505.

Maybe we should try adding `-j1` to cirrus(I remember `-j1` didn't work before, but might work now).

wolfgangwalther commented:
> @wolfgangwalther Hmm.. are you sure it's a timeout? It looks like it's failing because of low memory: [haskell/cabal#5505](https://github.com/haskell/cabal/issues/5505).

Absolutely, the low memory makes sense. I did not look carefully - I just figured it was not a real build error. I had 3 cirrus builds failing at 15 minutes, so I was a bit too quick to assume it was a timeout thing.

> **Edit**: The`-j1` seems to have worked. I added the change on #1600 (was also getting the cirrus error here).

Sweet, so I won't look at this here and just see green where red is :)

steve-chavez commented:
> The printed config is itself a valid postgrest config file.

I thought this would mean I could do `postgrest --dump-config > pgrst.conf` and generate a default config. But now I get it's mostly for tests. I'll give it a review now :)

steve-chavez commented on src/PostgREST/Config.hs:
```diff
@@ -225,6 +241,62 @@ readPathShowHelp = customExecParser parserPrefs opts
+  exitSuccess
+
+  where
+    dump = intercalate "\n" $ (\(k, v) -> k <> " = " <> v) <$>
```
I think [unlines](https://hackage.haskell.org/package/text-1.2.4.0/docs/Data-Text.html#v:unlines) could replace `intercalate` here?

wolfgangwalther commented:
> I thought this would mean I could do `postgrest --dump-config > pgrst.conf` and generate a default config. But now I get it's mostly for tests.

I think we're missing two things to be able to do that:
- the `FILENAME` argument must be optional. This is something that I have to change anyway to support "all-environment-only" configurations.
- with `--dump-config` enabled, postgrest should not throw when missing one of the required settings. This is something that could happen with the environment variables approach - not 100% yet. Depends on implementation of that.
- A good example config file needs comments, just like we provide them with the help output.

Maybe we can do that a bit different - what if we split the output of `--help` / the error output on stdout and stderr, so that only the example config file is on stdout and everything else on stderr? Then you should be able to do `postgrest --help > example.config` and have our example ready to go.

steve-chavez commented:
> what if we split the output of --help / the error output on stdout and stderr, so that only the example config file is on stdout and everything else on stderr? Then you should be able to do postgrest --help > example.config and have our example ready to go.

That sounds good! Seems like a simpler implementation.

wolfgangwalther commented:
> > what if we split the output of --help / the error output on stdout and stderr, so that only the example config file is on stdout and everything else on stderr? Then you should be able to do postgrest --help > example.config and have our example ready to go.
> 
> That sounds good! Seems like a simpler implementation.

Had a look at that.. and the implementation is not as easy as thought. The whole thing is dumped by `Options.Applicative`.

Implementation-wise we could just do the following:
- remove the example config from the options help output
- add another CLI flag `--dump-example-config` to print just that

I think that's also better to use. WDYT?

---

Anyway, this would be something for a different PR. I addressed the other comments, CI is passing -> merge time!
